### PR TITLE
Better default for drop_last in a distributed setting

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR sets a better default for `drop_last` in a distributed setting.

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--19478.org.readthedocs.build/en/19478/

### CODE REVIEW

File: src/lightning/data/streaming/dataset.py [hunk 1]
```diff
@@ -68,12 +73,24 @@ def __init__(
...
+        self.distributed_env = _DistributedEnv.detect()
+
+        if self.distributed_env.world_size > 1:
+            if drop_last is None or drop_last is True:
+                self.drop_last = True
+            else:
+                logger.warn(
+                    "You are running in a distributed environment and drop_last needs to be False. "
+                    "This can cause your training to hang if you are relying on distributed collectives."
+                )
```
Comment (ethanwharris) [inline src/lightning/data/streaming/dataset.py:85]: Language here is a bit strange - it says "drop_last needs to be False" but drop last is already False here? Also it then just carries on anyway so maybe it should instead override it and make drop_last true, or just throw an exception

File: src/lightning/data/streaming/dataset.py [hunk 2]
```diff
@@ -68,12 +73,24 @@ def __init__(
...
+                self.drop_last = drop_last
+        else:
+            self.drop_last = drop_last or False
```
Comment (rohitgr7) [inline src/lightning/data/streaming/dataset.py:88]:
```suggestion
        if self.distributed_env.world_size > 1:
            if drop_last is False:
                logger.warn(
                    "You are running in a distributed environment but `drop_last=False`. "
                    "This can cause your training to hang if you are relying on distributed collectives."
                )
            else:
                drop_last = True

        self.drop_last = drop_last or False
```

small change in warning message and can it be on rank zero?

**Review by rohitgr7 — COMMENTED** [review]
Thanks!

File: src/lightning/data/streaming/dataset.py (hunk 1 above)
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py]: Improve the messaging. It shouldn't raise an exception as this is correct for the validation set. However, it is important for users to be aware of potential issues.

File: src/lightning/data/streaming/dataset.py [hunk 3]
```diff
@@ -68,12 +73,24 @@ def __init__(
 
         self.item_loader = item_loader
         self.shuffle: bool = shuffle
-        self.drop_last = drop_last
+        self.distributed_env = _DistributedEnv.detect()
+
+        if self.distributed_env.world_size > 1:
+            if drop_last is False:
+                logger.warn(
+                    "You're operating within a distributed environment and have enabled the `drop_last` option. "
```
Comment (ethanwharris) [inline src/lightning/data/streaming/dataset.py:81]: Looks flipped again, now it says 'you have enable the `drop_last` option' when it is actually false

Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:81]: ChatGPT reverted it back ahah I should have verified instead of trusting it ;)

Comment (tchaton) [reply src/lightning/data/streaming/dataset.py]: done !

**Review by ethanwharris — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
